src/amoc_api_scenario_handler.erl

```diff
@@ -10,112 +10,125 @@
-            produces => ["application/json"]
-          },
-          patch =>
-          #{tags => ["scenario"],
-            description => "Starts scenario",
-            produces => ["application/json"]
-          }
-    },
-    [trails:trail("/scenarios/:id", amoc_api_scenario_handler, [], Metadata)].
+    #{get =>
+      #{tags => ["scenario"],
+        description => "Gets scenario status",
+        produces => ["application/json"]
+       },
+      patch =>
+      #{tags => ["scenario"],
+        description => "Starts scenario",
+        produces => ["application/json"]
+       }
+     },
+    [trails:trail("/scenarios/:id", ?MODULE, [], Metadata)].
 
 -spec init(tuple(), cowboy:req(), state()) -> {upgrade, protocol, cowboy_rest}.
 init({tcp, http}, _Req, _Opts) ->
     {upgrade, protocol, cowboy_rest}.
 
 -spec rest_init(cowboy:req(), [atom()]) -> {ok, cowboy:req(), state()}.
-rest_init(Req, [Action]) ->
-    {ok, Req, #state{action = Action}}.
+rest_init(Req, _Opts) ->
+    {PathB, Req2} = cowboy_req:path(Req),
+    Resource = get_resource_string(PathB),
+    {ok, Req2, #state{resource = Resource}}.
 
 -spec allowed_methods(cowboy:req(), state()) -> 
-        {[binary()], cowboy:req(), state()}.
+    {[binary()], cowboy:req(), state()}.
 allowed_methods(Req, State) ->
     {[<<"PATCH">>, <<"GET">>], Req, State}.
 
 -spec content_types_provided(cowboy:req(), state()) -> 
-        {[tuple()], cowboy:req(), state()}.
+    {[tuple()], cowboy:req(), state()}.
 content_types_provided(Req, State) ->
-    {[{<<"application/json">>, process_request}], Req, State}.
+    {[{<<"application/json">>, handle_get}], Req, State}.
 
 -spec content_types_accepted(cowboy:req(), state()) -> 
-        {[tuple()], cowboy:req(), state()}.
+    {[tuple()], cowboy:req(), state()}.
 content_types_accepted(Req, State) ->
-    {[{<<"application/json">>, process_request}], Req, State}.
-
--spec process_request(cowboy:req(), state()) -> 
-    {halt, cowboy:req(), state()}.
-process_request(Req0, State) ->
-    {ok, Data, Req1} = cowboy_req:body(Req0),
-    ContentType = {<<"content-type">>, <<"application/json">>},
-    try
-        Term = case jsx:is_json(Data) of
-            true -> jsx:decode(Data);
-            false -> undef
-        end,
-        {Status, Reply} = process_json(Term, State),
-        lager:info(Reply),
-        {ok, Req2} = cowboy_req:reply(Status, [ContentType], Reply, Req1),
-        {halt, Req2, State}
-    catch _:_ ->
-        ReplyE = jsx:encode([{<<"error">>, <<"unknown">>}]),
-        {ok, ReqE} = cowboy_req:reply(500, [ContentType], ReplyE, Req1),
-        {halt, ReqE, State}
+    {[{<<"application/json">>, handle_patch}], Req, State}.
+
+-spec malformed_request(cowboy:req(), state()) ->
+    {boolean(), cowboy:req(), state()}.
+malformed_request(Req, State) ->
+    {Method, Req2} = cowboy_req:method(Req),
+    malformed_request(Method, Req2, State).
+
+malformed_request(<<"PATCH">>, Req, State) ->
```

Comment:
This callback is not expected to deal with body. According to the [cowboy flowcharts](http://ninenines.eu/docs/en/cowboy/2.0/guide/rest_flowcharts/) this checks is expected to deal only with the  request line and request headers.

> A series of callbacks are called in succession to perform a general checkup of the service, the request line and request headers.

> The request body, if any, is not expected to have been received for any of these steps. It is only processed at the end of the "PUT, POST and PATCH methods" diagram, when all conditions have been met.